luna.c

```diff
@@ -121,34 +121,23 @@ void *escape_unicode(char *in_buf, size_t header_size, size_t footer_size, size_
 	return out_buf2;
 }
 
-/* sub-routine of xml_compress() to escape the Lua script string. Returns the new in_buf or NULL */
-void *escape_special_xml_chars(char *in_buf, size_t header_size, size_t in_size, size_t *obuf_size) {
-	char *p;
-	unsigned extend_with = 0;
-	for (p = in_buf + header_size; p < in_buf + header_size + in_size; p++) {
-		if (*p == '&') extend_with += 4; // amp;
-		else if (*p == '<') extend_with += 3; // lt;
-	}
-	if (extend_with) {
-		*obuf_size += extend_with;
-		void *tmp_in_buf;
-		if (!(tmp_in_buf = realloc(in_buf, *obuf_size))) {
-			puts("can't realloc in_buf for special characters");
-			free(in_buf);
-			return NULL;
-		}
-		in_buf = tmp_in_buf;
-		unsigned new_written = 0;
-		for (p = in_buf + header_size; p < in_buf + header_size + in_size + new_written; p++) {
-			if (*p == '&') {
-				memmove(p + 5, p + 1, in_buf + header_size + in_size + new_written - p - 1);
-				memcpy(p, "&amp;", 5);
-				new_written += 4;
-			} else if (*p == '<') {
-				memmove(p + 4, p + 1, in_buf + header_size + in_size + new_written - p - 1);
-				memcpy(p, "&lt;", 4);
-				new_written += 3;
+/* sub-routine of xml_compress() to fix occurrences of CDATA end sequence "]]>" in Lua scripts
+ * by spitting them between two CDATA sections. Returns the new in_buf, or NULL if out of memory. */
+void *fix_cdata_end_seq(char *in_buf, size_t header_size, size_t in_size, size_t *obuf_size) {
+	for (size_t offset = header_size; offset < header_size + in_size - 2; offset++) {
+		if (!memcmp(in_buf + offset, "]]>", 3)) {
+			*obuf_size += 12; // ]]><![CDATA[
+			char *tmp_in_buf;
+			if (!(tmp_in_buf = realloc(in_buf, *obuf_size))) {
+				puts("can't realloc in_buf");
+				free(in_buf);
+				return NULL;
 			}
+			in_buf = tmp_in_buf;
+			memmove(in_buf + offset + 14, in_buf + offset + 2, header_size + in_size - (offset + 2));
+			memcpy(in_buf + offset + 2, "]]><![CDATA[", 12);
+			in_size += 12;
+			offset += 14;
```

Comment:
This might have to be offset + 13, to account for the `offset++` every iteration. Could be verified by using `]]>]]>` as input.

Reply:
Nope, the additional `offset++` actually skips past the `>`.